File: app/services/customer_service.py
"""
客户服务层
"""

from sqlalchemy.orm import Session
from sqlalchemy import and_, or_
from typing import List, Optional, Tuple
import math
from datetime import datetime

from ..models.customer import (
    Customer, CustomerCreate, CustomerUpdate, CustomerProgressUpdate, CustomerEmailCountUpdate,
    CommunicationProgress, InterestLevel
)


class CustomerService:
    """客户服务类"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_customer(self, customer_data: CustomerCreate, user_id: int) -> Customer:
        """创建客户"""
        db_customer = Customer(
            user_id=user_id,
            name=customer_data.name,
            email=customer_data.email,
            company=customer_data.company,
            email_count=customer_data.email_count,
            communication_progress=customer_data.communication_progress,
            interest_level=customer_data.interest_level,
            last_communication_time=customer_data.last_communication_time,
            current_progress=customer_data.current_progress
        )
        self.db.add(db_customer)
        self.db.commit()
        self.db.refresh(db_customer)
        return db_customer
    
    def get_customer(self, customer_id: int, user_id: int) -> Optional[Customer]:
        """获取单个客户"""
        return self.db.query(Customer).filter(
            and_(Customer.id == customer_id, Customer.user_id == user_id)
        ).first()
    
    def get_customers(
        self, 
        user_id: int, 
        page: int = 1, 
        page_size: int = 20,
        search_query: Optional[str] = None,
        communication_progress: Optional[CommunicationProgress] = None,
        interest_level: Optional[InterestLevel] = None
    ) -> Tuple[List[Customer], int]:
        """获取客户列表"""
        query = self.db.query(Customer).filter(Customer.user_id == user_id)
        
        # 搜索功能
        if search_query:
            search_filter = or_(
                Customer.name.ilike(f"%{search_query}%"),
                Customer.email.ilike(f"%{search_query}%"),
                Customer.company.ilike(f"%{search_query}%")
            )
            query = query.filter(search_filter)
        
        # 沟通进度筛选
        if communication_progress:
            query = query.filter(Customer.communication_progress == communication_progress)
        
        # 感兴趣程度筛选
        if interest_level:
            query = query.filter(Customer.interest_level == interest_level)
        
        # 按创建时间倒序排序
        query = query.order_by(Customer.created_at.desc())
        
        # 计算总数
        total = query.count()
        
        # 分页
        offset = (page - 1) * page_size
        customers = query.offset(offset).limit(page_size).all()
        
        return customers, total
    
    def update_customer(self, customer_id: int, customer_data: CustomerUpdate, user_id: int) -> Optional[Customer]:
        """更新客户信息"""
        db_customer = self.get_customer(customer_id, user_id)
        if not db_customer:
            return None
        
        update_data = customer_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_customer, field, value)
        
        self.db.commit()
        self.db.refresh(db_customer)
        return db_customer
    
    def update_customer_progress(self, customer_id: int, progress_data: CustomerProgressUpdate, user_id: int) -> Optional[Customer]:
        """更新客户进度"""
        db_customer = self.get_customer(customer_id, user_id)
        if not db_customer:
            return None
        
        update_data = progress_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_customer, field, value)
        
        self.db.commit()
        self.db.refresh(db_customer)
        return db_customer
    
    def update_email_count(self, customer_id: int, email_data: CustomerEmailCountUpdate, user_id: int) -> Optional[Customer]:
        """更新邮件计数"""
        db_customer = self.get_customer(customer_id, user_id)
        if not db_customer:
            return None
        
        db_customer.email_count = email_data.email_count
        if email_data.last_communication_time:
            db_customer.last_communication_time = email_data.last_communication_time
        else:
            db_customer.last_communication_time = datetime.now()
        
        self.db.commit()
        self.db.refresh(db_customer)
        return db_customer
    
    def delete_customer(self, customer_id: int, user_id: int) -> bool:
        """删除客户"""
        db_customer = self.get_customer(customer_id, user_id)
        if not db_customer:
            return False
        
        self.db.delete(db_customer)
        self.db.commit()
        return True
    
    def get_customers_by_progress(self, user_id: int, progress: CommunicationProgress) -> List[Customer]:
        """根据沟通进度获取客户列表"""
        return self.db.query(Customer).filter(
            and_(Customer.user_id == user_id, Customer.communication_progress == progress)
        ).all()
    
    def get_customers_by_interest(self, user_id: int, interest: InterestLevel) -> List[Customer]:
        """根据感兴趣程度获取客户列表"""
        return self.db.query(Customer).filter(
            and_(Customer.user_id == user_id, Customer.interest_level == interest)
        ).all()
    
    def get_customer_statistics(self, user_id: int) -> dict:
        """获取客户统计信息"""
        total_customers = self.db.query(Customer).filter(Customer.user_id == user_id).count()
        
        # 按沟通进度统计
        progress_stats = {}
        for progress in CommunicationProgress:
            count = self.db.query(Customer).filter(
                and_(Customer.user_id == user_id, Customer.communication_progress == progress)
            ).count()
            progress_stats[progress.value] = count
        
        # 按感兴趣程度统计
        interest_stats = {}
        for interest in InterestLevel:
            count = self.db.query(Customer).filter(
                and_(Customer.user_id == user_id, Customer.interest_level == interest)
            ).count()
            interest_stats[interest.value] = count
        
        return {
            "total_customers": total_customers,
            "communication_progress": progress_stats,
            "interest_level": interest_stats
        }

File: hnode/src/services/CustomerService.js
const { Op } = require('sequelize');
const { Customer, Contract } = require('../models');
const { sequelize } = require('../config/database');

class CustomerService {
  constructor() {}

  async createCustomer(customerData, userId) {
    try {
      const customer = await Customer.create({
        ...customerData,
        user_id: userId
      });
      return customer;
    } catch (error) {
      throw new Error(`创建客户失败: ${error.message}`);
    }
  }

  async getCustomer(customerId, userId) {
    try {
      const customer = await Customer.findOne({
        where: {
          id: customerId,
          user_id: userId
        }
      });
      return customer;
    } catch (error) {
      throw new Error(`获取客户失败: ${error.message}`);
    }
  }

  async getCustomers(userId, options = {}) {
    try {
      const {
        page = 1,
        pageSize = 20,
        search = null,
        communicationProgress = null,
        interestLevel = null
      } = options;

      // 构建SQL WHERE条件
      let sqlWhere = `user_id = ${userId}`;
      const replacements = [];
      
      if (search) {
        sqlWhere += ` AND (name ILIKE $1 OR first_name ILIKE $1 OR last_name ILIKE $1 OR email ILIKE $1 OR company ILIKE $1)`;
        replacements.push(`%${search}%`);
      }
      
      if (communicationProgress) {
        sqlWhere += ` AND communication_progress = $${replacements.length + 1}`;
        replacements.push(communicationProgress);
      }
      
      if (interestLevel) {
        sqlWhere += ` AND interest_level = $${replacements.length + 1}`;
        replacements.push(interestLevel);
      }
      
      // 使用原始SQL查询（完全绕过Sequelize模型）
      const offset = (page - 1) * pageSize;
      
      console.log('📊 执行SQL查询，WHERE条件:', sqlWhere);
      
      // 获取总数
      const [countResult] = await sequelize.query(
        `SELECT COUNT(*) as total FROM customers WHERE ${sqlWhere}`,
        { 
          bind: replacements,
          raw: true 
        }
      );
      const total = parseInt(countResult[0].total);
      
      // 获取客户数据（包含deal_status）
      const [customers] = await sequelize.query(
        `SELECT 
          id, user_id, name, first_name, last_name, email, company,
          email_count, communication_progress, interest_level, deal_status,
          last_communication_time, created_at, updated_at
         FROM customers 
         WHERE ${sqlWhere}
         ORDER BY created_at DESC
         LIMIT ${pageSize} OFFSET ${offset}`,
        { 
          bind: replacements,
          raw: true 
        }
      );
      
      console.log('✅ 原始SQL查询完成，共', customers.length, '条记录');
      if (customers.length > 0) {
        console.log('🔍 第一个客户数据:', customers[0]);
      }
      
      // 为每个客户添加合同数量
      const customersWithCount = await Promise.all(customers.map(async (customer) => {
        try {
          const contractCount = await Contract.count({ where: { customer_id: customer.id } });
          customer.contract_count = contractCount;
        } catch (error) {
          customer.contract_count = 0;
        }
        return customer;
      }));

      return {
        customers: customersWithCount,
        total: total,
        page,
        pageSize,
        totalPages: Math.ceil(total / pageSize)
      };
    } catch (error) {
      throw new Error(`获取客户列表失败: ${error.message}`);
    }
  }

  async updateCustomer(customerId, customerData, userId) {
    try {
      const [updatedRowsCount] = await Customer.update(customerData, {
        where: {
          id: customerId,
          user_id: userId
        }
      });

      if (updatedRowsCount === 0) {
        return null;
      }

      return await this.getCustomer(customerId, userId);
    } catch (error) {
      throw new Error(`更新客户失败: ${error.message}`);
    }
  }

  async updateCustomerProgress(customerId, progress, userId) {
    try {
      const validProgresses = ['待联系', '跟进中', '不再跟进', '暂停跟进'];
      if (!validProgresses.includes(progress)) {
        throw new Error('无效的沟通进度');
      }

      const [updatedRowsCount] = await Customer.update(
        { communication_progress: progress },
        {
          where: {
            id: customerId,
            user_id: userId
          }
        }
      );

      if (updatedRowsCount === 0) {
        return null;
      }

      return await this.getCustomer(customerId, userId);
    } catch (error) {
      throw new Error(`更新客户进度失败: ${error.message}`);
    }
  }

  async updateCustomerInterestLevel(customerId, interestLevel, userId) {
    try {
      const validLevels = ['无兴趣', '低兴趣', '中等兴趣', '高兴趣'];
      if (!validLevels.includes(interestLevel)) {
        throw new Error('无效的兴趣程度');
      }

      const [updatedRowsCount] = await Customer.update(
        { interest_level: interestLevel },
        {
          where: {
            id: customerId,
            user_id: userId
          }
        }
      );

      if (updatedRowsCount === 0) {
        return null;
      }

      return await this.getCustomer(customerId, userId);
    } catch (error) {
      throw new Error(`更新客户兴趣程度失败: ${error.message}`);
    }
  }

  async deleteCustomer(customerId, userId) {
    try {
      const deletedRowsCount = await Customer.destroy({
        where: {
          id: customerId,
          user_id: userId
        }
      });

      return deletedRowsCount > 0;
    } catch (error) {
      throw new Error(`删除客户失败: ${error.message}`);
    }
  }

  async getCustomerStatistics(userId) {
    try {
      const totalCustomers = await Customer.count({
        where: { user_id: userId }
      });

      // 按沟通进度统计
      const progressStats = await Customer.findAll({
        attributes: [
          'communication_progress',
          [require('sequelize').fn('COUNT', require('sequelize').col('id')), 'count']
        ],
        where: { user_id: userId },
        group: ['communication_progress']
      });

      // 按兴趣程度统计
      const interestStats = await Customer.findAll({
        attributes: [
          'interest_level',
          [require('sequelize').fn('COUNT', require('sequelize').col('id')), 'count']
        ],
        where: { user_id: userId },
        group: ['interest_level']
      });

      // 最近沟通的客户数量
      const recentCommunicationCount = await Customer.count({
        where: {
          user_id: userId,
          last_communication_time: {
            [Op.gte]: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000) // 最近7天
          }
        }
      });

      return {
        total_customers: totalCustomers,
        progress_statistics: progressStats.reduce((acc, item) => {
          acc[item.communication_progress] = parseInt(item.dataValues.count);
          return acc;
        }, {}),
        interest_statistics: interestStats.reduce((acc, item) => {
          acc[item.interest_level] = parseInt(item.dataValues.count);
          return acc;
        }, {}),
        recent_communication_count: recentCommunicationCount
      };
    } catch (error) {
      throw new Error(`获取客户统计失败: ${error.message}`);
    }
  }
}

module.exports = CustomerService;
